### PR TITLE
stat_boxplot warns if x is continuous and no group aesthetics is specified

### LINKED ISSUE #992: geom_boxplot with continuous x and without group aesthetics
```
d <- data.frame(x=rep(1:10, 100), y=runif(1000))
ggplot(d) + geom_boxplot(aes(x=x, y=y))
```

will happily produce a single box, while the intention is probably to produce the output of this code:

```
d <- data.frame(x=rep(1:10, 100), y=runif(1000))
ggplot(d) + geom_boxplot(aes(x=x, y=y, group=x))
```

Would a warning be appropriate here -- if `x` is continuous and the `group` aesthetics is missing?

### PR DESCRIPTION
**Pull request by krlmlr**
- Only if there is more than one x value
- With tests

Fixes #992.

Closes #1213 (=includes it).

### CODE REVIEW

File: R/stat-boxplot.r [hunk 1]
```diff
@@ -45,6 +45,11 @@ StatBoxplot <- ggproto("StatBoxplot", Stat,
   {
     data <- remove_missing(data, na.rm, c("x", "y", "weight"), name="stat_boxplot",
       finite = TRUE)
+    if (!is.discrete(data$x) && data$group[1L] == NO_GROUP && any(data$x != data$x[1L])) {
```
Comment (hadley) [inline R/stat-boxplot.r]: Did you check this? I think by this point factors have been converted into integers, so the appropriate test is `is.numeric(data$x)`.  And I think `all(data$group == NO_GROUP)` would be clearer

**Comment by hadley** [comment]
Could you please squash the commits?

File: R/stat-boxplot.r (hunk 1 above)
Comment (krlmlr) [inline R/stat-boxplot.r]: Thanks. I'll add a test later tonight.

`data$group[1L] == NO_GROUP` is much faster for the (expected) case that the user has specified a group. I'd rather not add an extra slowdown, instead I could add a comment or refactor this into a function. Please confirm.

Comment (hadley) [inline R/stat-boxplot.r]: Seems unlikely to be a bottleneck - easily understood code would be better

**Comment by krlmlr** [comment]
Squashed commits, all tests pass locally. I hope we can agree on `has_groups()` (#1213).

**Comment by krlmlr** [comment]
The case of facets hasn't been covered yet. I've added a test and included #1236 in this pull request.

File: R/stat-boxplot.r [hunk 2]
```diff
@@ -47,6 +47,11 @@ StatBoxplot <- ggproto("StatBoxplot", Stat,
   {
     data <- remove_missing(data, na.rm, c("x", "y", "weight"), name = "stat_boxplot",
       finite = TRUE)
+    if (!is.discrete(data$x) && !has_groups(data) && any(data$x != data$x[1L])) {
```
Comment (hadley) [inline R/stat-boxplot.r]: I'm pretty sure the correct test is `is.double()`

**Comment by krlmlr** [comment]
Changed test to `is.double()`. Tests pass locally.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**